Mathlib/Analysis/Complex/Hadamard.lean

```diff
@@ -0,0 +1,488 @@
+    left; simpa using ne_of_gt h0
+  · apply Differentiable.const_cpow (Differentiable.neg differentiable_id')
+    simp only [ne_eq, ofReal_eq_zero, neg_eq_zero, Or.inl (ne_of_gt h1)]
+
+/-- The function `f` is bounded by `sSupAbsIm`. (`f` is bounded on its domain.) -/
+lemma abs_le_sSupAbsIm (z : ℂ) (hD : z ∈ (closedStrip 0 1))
+    (hB : BddAbove ((norm ∘ f) '' (closedStrip 0 1))) : norm (f z) ≤ sSupAbsIm f (z.re) := by
+  refine le_csSup ?_ ?_
+  · apply BddAbove.mono (image_subset (norm ∘ f) _) hB
+    exact preimage_mono (singleton_subset_iff.mpr hD)
+  · apply mem_image_of_mem (norm ∘ f)
+    simp only [mem_preimage, mem_singleton]
+
+/-- The function `F` is bounded above because `f` is. -/
+lemma F_BddAbove (hB : BddAbove ((norm ∘ f) '' (closedStrip 0 1))) :
+    BddAbove ((norm ∘ (F f)) '' (closedStrip 0 1)) := by
+ -- Rewriting goal
+  simp only [F, image_congr, comp_apply, map_mul, invInterpStrip]
+  rw [bddAbove_def] at *
+  cases' hB with B hB
+  -- Using bound
+  use ((max 1 ((sSupAbsIm f 0)^ (-(1 : ℝ)))) * max 1 ((sSupAbsIm f 1)^ (-(1 : ℝ)))) * B
+  simp only [mem_image, forall_exists_index, and_imp, forall_apply_eq_imp_iff₂]
+  intros z hset
+  specialize hB (norm (f z))
+  specialize hB _
+  · simp only [image_congr, mem_image, comp_apply]; use z
+  -- Proof that the bound is correct
+  · rw [norm_smul]
+    apply mul_le_mul _ hB (norm_nonneg _) _
+    · rw [norm_mul]
+      apply mul_le_mul _ _ (norm_nonneg _) (le_trans zero_le_one (le_max_left _ _))
+      -- Bounding individual terms
+      · by_cases hM0_one : 1 ≤ sSupAbsIm f 0
+        -- `1 ≤ (sSupAbsIm f 0)`
+        · apply le_trans _ (le_max_left _ _)
+          simp only [norm_eq_abs, abs_cpow_eq_rpow_re_of_pos h0, sub_re, one_re,
+          Real.rpow_le_one_of_one_le_of_nonpos hM0_one (sub_nonpos.mpr hset.2)]
+        -- `0 < (sSupAbsIm f 0) < 1`
+        · rw [not_le] at hM0_one; apply le_trans _ (le_max_right _ _)
+          simp only [norm_eq_abs, abs_cpow_eq_rpow_re_of_pos h0, sub_re, one_re]
+          apply Real.rpow_le_rpow_of_exponent_ge (h0) (le_of_lt hM0_one) _
+          simp only [neg_le_sub_iff_le_add, le_add_iff_nonneg_left, hset.1]
+      · by_cases hM1_one : 1 ≤ sSupAbsIm f 1
+        -- `1 ≤ sSupAbsIm f 1`
+        · apply le_trans _ (le_max_left _ _)
+          simp only [norm_eq_abs, abs_cpow_eq_rpow_re_of_pos h1, sub_re, one_re, neg_re,
+          Real.rpow_le_one_of_one_le_of_nonpos hM1_one (Right.neg_nonpos_iff.mpr hset.1)]
+        -- `0 < sSupAbsIm f 1 < 1`
+        · rw [not_le] at hM1_one; apply le_trans _ (le_max_right _ _)
+          simp only [norm_eq_abs, abs_cpow_eq_rpow_re_of_pos h1, sub_re, one_re, neg_re,
+          Real.rpow_le_rpow_of_exponent_ge (h1) (le_of_lt hM1_one)
+            (neg_le_neg_iff.mpr hset.2)]
+    simp only [gt_iff_lt, lt_max_iff, zero_lt_one, true_or, mul_nonneg_iff_of_pos_left, le_max_iff,
+      zero_le_one]
+
+-- Proof that the edges are bounded by one
```

Comment:
Can you give a true docstring, saying what the lemma says and under which assumptions?